src/NHibernate/Transaction/AdoTransaction.cs

```diff
@@ -190,20 +190,19 @@ public void Commit()
 
 				log.Debug("Start Commit");
 
+				//Hibernate calls flush from inside BeforeTransactionCompletion
+				if (session.FlushMode != FlushMode.Manual)
+					session.Flush();
+				session.BeforeTransactionCompletion(this);
```

Comment:
`ShouldFlushFromDistributedTransaction`? No more existing I believe. There is only the `using FlushingFromDtcTransaction`, which would have to surround `BeforeTransactionCompletion` instead of flush. That would by the way remove the inconsistency in system transactions, which flushes after `BeforeTransactionCompletion` instead of before like `AdoTransaction`. It does not look to me there is any reason for this inconsistency.
There is the test on whether the connection manager is still connected: it could surround all. If it is no more connected, that means the session is no more participating in the transaction (session closed very likely), no event should be raised for it. Or we could handle that inside the `BeforeTransactionCompletion` if we still want to notify closed sessions for transaction in which there were enlisted.